src/common/greptimedb-telemetry/src/lib.rs

```diff
@@ -0,0 +1,283 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::env;
+use std::io::ErrorKind;
+use std::time::Duration;
+
+use common_runtime::error::{Error, Result};
+use common_runtime::{BoxedTaskFunction, RepeatedTask, Runtime, TaskFunction};
+use common_telemetry::debug;
+use reqwest::{Client, Response};
+use serde::{Deserialize, Serialize};
+
+pub const TELEMETRY_URL: &str = "https://api-preview.greptime.cloud/db/otel/statistics";
+const TELEMETRY_UUID_FILE_NAME: &str = "/tmp/.greptimedb_telemetry_uuid";
+
+pub static TELEMETRY_INTERVAL: Duration = Duration::from_secs(60 * 30);
+
+//pub type GreptimeDBTelemetryTask = RepeatedTask<Error>;
+
+pub enum GreptimeDBTelemetryTask {
+    Enable(RepeatedTask<Error>),
+    Disable,
+}
+
+impl GreptimeDBTelemetryTask {
+    pub fn enable(interval: Duration, task_fn: BoxedTaskFunction<Error>) -> Self {
+        GreptimeDBTelemetryTask::Enable(RepeatedTask::new(interval, task_fn))
+    }
+
+    pub fn disable() -> Self {
+        GreptimeDBTelemetryTask::Disable
+    }
+
+    pub fn start(&self, runtime: Runtime) -> Result<()> {
+        match self {
+            GreptimeDBTelemetryTask::Enable(task) => task.start(runtime),
+            GreptimeDBTelemetryTask::Disable => Ok(()),
+        }
+    }
+
+    pub async fn stop(&self) -> Result<()> {
+        match self {
+            GreptimeDBTelemetryTask::Enable(task) => task.stop().await,
+            GreptimeDBTelemetryTask::Disable => Ok(()),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct StatisticData {
+    pub os: String,
+    pub version: String,
+    pub arch: String,
+    pub mode: Mode,
+    pub git_commit: String,
+    pub nodes: Option<i32>,
+    pub uuid: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Eq, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum Mode {
+    Distributed,
+    Standalone,
+}
+
+#[async_trait::async_trait]
+pub trait Collector {
+    fn get_version(&self) -> String {
+        env!("CARGO_PKG_VERSION").to_string()
+    }
+
+    fn get_git_hash(&self) -> String {
+        env!("GIT_COMMIT").to_string()
+    }
+
+    fn get_os(&self) -> String {
+        env::consts::OS.to_string()
+    }
+
+    fn get_arch(&self) -> String {
+        env::consts::ARCH.to_string()
+    }
+
+    fn get_mode(&self) -> Mode;
+
+    fn get_retry(&self) -> i32;
+
+    fn inc_retry(&mut self);
```

Comment:
I don't find any code calling the `inc_retry` function. Do i miss something?

Reply:
Good catch. I will fix it.